chessengine/utils.py

```diff
@@ -75,3 +76,33 @@ def clear_lines(n: int) -> None:
     LINE_CLEAR = "\x1b[2K"
     for i in range(n):
         print(LINE_UP, end=LINE_CLEAR)
+
+
+def to_san(move: str) -> str:
```

Comment:
I had a slightly different idea of how the `to_san` function would work. Right now it allows the user to be a little more lax with their input, but it still needs them to kind of be familiar with SAN. 

I imagined the `to_san` function to take a move that was not described in SAN, but some other (much more straightforward) format that we would document and tell the users to use if they don't know SAN, and convert it into valid SAN.

For example, we could document that if users don't know SAN, they could enter their moves in the format - `square1 to square2` (e.g. `b1 to c3`), and the `to_san` function would convert it into valid SAN (e.g. `Nc3`). To do this we would need to identify the pieces on the start and end square, which can be done using `Board.identify_piece_at`, and then generating the SAN move wouldn't be too hard. That's what I had in mind for the `to_san` function.

Reply:
> I had a slightly different idea of how the `to_san` function would work. Right now it allows the user to be a little more lax with their input, but it still needs them to kind of be familiar with SAN.
> 
> I imagined the `to_san` function to take a move that was not described in SAN, but some other (much more straightforward) format that we would document and tell the users to use if they don't know SAN, and convert it into valid SAN.
> 
> For example, we could document that if users don't know SAN, they could enter their moves in the format - `square1 to square2` (e.g. `b1 to c3`), and the `to_san` function would convert it into valid SAN (e.g. `Nc3`). To do this we would need to identify the pieces on the start and end square, which can be done using `Board.identify_piece_at`, and then generating the SAN move wouldn't be too hard. That's what I had in mind for the `to_san` function.

Ah, alright, I guess in that case it indeed makes sense to define it inside the `Board` class. (As we need access to the board object in order to call `identify_piece_at` on it.)

Will work on it. I guess this is so a chance to get a bit more familiar with the internals of the `Board` class. :)

Reply:
> I'd rather say, we should refactor the #move_san function in the following way:
> Have one function that - given start and destination coordinates (or the corresponding index in the bitboard) - just checks if the move is legal and executes it.
> And then have a wrapper around this that translates a move from SAN format into coordinates (or into the index on the bitboard directly).

This is already how `move_san` works :) The `Board.move` function accepts start and end positions and makes the move if it is valid. `move_san` then takes the SAN move, extracts start and end positions from it, and calls `move` on those positions.

> I'm not sure if it's a good idea to translate the user input into SAN if we already have it in a much clearer format.

That's true, but the reason I thought a `to_san` function/method would help was because then we could call it on non-SAN moves just before passing it to `move_san`, and we wouldn't have to add any more code to the `play` and `play_pvp` methods. In case we don't add a `to_san` method and try to extract the start and end positions after we get the user input, we would have to add another conditional statement to the `play` methods ([possibly here](https://github.com/hrushikeshrv/chessengine/blob/main/chessengine/bitboard.py#L682)), which would make the `play` methods even longer.

Reply:
> This is already how `move_san` works :) The `Board.move` function accepts start and end positions and makes the move if it is valid. `move_san` then takes the SAN move, extracts start and end positions from it, and calls `move` on those positions.

As far as I understand form looking at the code, it's only the `move_san` method that checks if the move is valid. (When it calls `self.get_moves` in what is currently line 440.)

The `move` function, as far as I can see, just executes any move (as long as the coordinates stay within the bounds of the board). Am I seeing this wrong?

So, basically, we have 3 "tasks" that these functions attempt to do:  
1. parse the input format (so far, it's always SAN) into the `coordinates` format that the internal board works with
2. check whether the given move is legal
3. execute it by updating the internal state accordingly

Now currently, if I understand the code correctly, `move_san` method does 1. & 2., whereas the `move` method does 3.  
In order to allow multiple input formats, I think we should separate tasks 1. & 2., i.e. have them handled by separate methods. Then we could easily add additional methods for handling new input formats.

Using SAN as a kind of common intermediate format that all other formats get translated to is not a good idea imo.  
The problem is that SAN, or rather translating something into SAN, is in fact really complicated, because of all those weird edge cases, ambiguities (like two knights/rooks being able to move to the same square).  
I think an intermediate format (that all inputs get parsed to) should be as simple as possible / as close to the machine's representation as possible.

Hope this makes it clear what I meant. ^^

Reply:
> As far as I understand form looking at the code, it's only the move_san method that checks if the move is valid.

You're right, I misspoke. I agree with your reasoning. We should proceed according to what you said, i.e. separating tasks 1 & 2.

Reply:
Alright, will start working on it.